### PR TITLE
feat(*): Kyberswap Elastic integration

### PR DESCRIPTION
**Pull request by linhbkhn95**
### Why did we need it?
- Integrate KyberSwap Elastic
   + Passed tests for Simple Swap (SELL, BUY)

### CODE REVIEW

File: src/dex/ks-elastic/ks-elastic-pool.ts [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+  async generateState(blockNumber: number): Promise<Readonly<PoolState>> {
+    // TODO: Should be handle it first before processing other logics
+    // Cache pool contract for next process
+    if (!this.isSetPoolContract) {
+      this.poolContract = await this.gePoolContract();
+      this.isSetPoolContract = true;
+    }
+
+    const batchRequestData = [
+      this.getAllTick(this.poolAddress, blockNumber),
```
Comment (adedade) [inline src/dex/ks-elastic/ks-elastic-pool.ts:298]: Should we cache these somewhere?

File: src/dex/ks-elastic/ks-elastic-pool.ts [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+      liquidity: bigIntify(_liquidityState.baseL),
+      ticks,
+      isValid,
+      currentTick: bigIntify(currentTick),
+      reinvestLiquidity: bigIntify(_liquidityState.reinvestL),
+    };
+  }
+
+  handleSwapEvent(event: any, pool: PoolState, log: Log) {
+    const newSqrtPriceX96 = bigIntify(event.args.sqrtPriceX96);
```
Comment (adedade) [inline src/dex/ks-elastic/ks-elastic-pool.ts:334]: These values seem wrong @linhbkhn95, I believe the correct version is:

```ts
const newSqrtPriceX96 = bigIntify(event.args.sqrtP);
const amount0 = bigIntify(event.args.deltaQty0);
const newTick = bigIntify(event.args.currentTick);
const newLiquidity = bigIntify(event.args.liquidity);
```

File: src/dex/ks-elastic/ks-elastic-pool.ts [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+        tickLower,
+        tickUpper,
+        liquidityDelta: -BigInt.asIntN(128, BigInt.asIntN(256, amount)),
+      }),
+      pool,
+    );
+  }
+
+  handleMintEvent(event: any, pool: PoolState, log: Log) {
+    const amount = bigIntify(event.args.amount);
```
Comment (adedade) [inline src/dex/ks-elastic/ks-elastic-pool.ts:380]: Should be

```ts
const amount = bigIntify(event.args.qty);
```

File: src/dex/ks-elastic/ks-elastic-pool.ts [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+            newLiquidity,
+            zeroForOne,
+          ),
+        pool,
+      );
+    }
+  }
+
+  handleBurnEvent(event: any, pool: PoolState, log: Log) {
+    const amount = bigIntify(event.args.amount);
```
Comment (adedade) [inline src/dex/ks-elastic/ks-elastic-pool.ts:365]: Same as https://github.com/paraswap/paraswap-dex-lib/pull/227/files/f04fbb3fef05e73b811088738f7d367c8c586bf0#r1046177799

File: src/dex/ks-elastic/ks-elastic-pool.ts [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+
+  getPoolState(blockNumber: number) {
+    const callRequest = this.getPoolStateRequest();
+    return this.poolContract.methods[callRequest.funcName](
+      ...callRequest.params,
+    ).call({}, blockNumber || 'latest');
+  }
+
+  getAllTick(poolAddress: string, blockNumber: number) {
+    const callRequest = this.getAllTickRequest(poolAddress);
```
Comment (adedade) [inline src/dex/ks-elastic/ks-elastic-pool.ts:253]: This function returns an out of gas error when running for top pools (I am assuming with lots of ticks). e.g. pool with address `0x087abaab9cd85025a8b3916948c69fe173c837ea` on Arbitrum

**Comment by adedade** [comment]
Hey @linhbkhn95 I did some testing on Arbitrum, and I am finding weird results when comparing your implementation with a an online quote (e.g. https://aggregator-api.kyberswap.com/arbitrum/route/encode?dexes=kyberswapv2). 

Below result is correct:
```
[2022-12-12T19:07:20.023] [DEBUG] dex - kyberswapElastic 0x114de2affc6a335433dbe9d3d51a8f31a5591fdf inSync: {
  fetched: { amount: '1.29180897503046588 wstETH', duration: '382ms' },
  offChain: { amount: '1.291812211653161226 wstETH', duration: '0ms' }
}
```

Below result returns always 0.0:
```
[2022-12-12T19:08:49.557] [DEBUG] dex - kyberswapElastic 0x114de2affc6a335433dbe9d3d51a8f31a5591fdf inSync: {
  fetched: { amount: '0.000999800008908787 WETH', duration: '459ms' },
  offChain: { amount: '0.0 WETH', duration: '1ms' }
}
```

Below result the pool is completely off (or returns 0.0:
```
[2022-12-12T19:07:09.906] [DEBUG] dex - kyberswapElastic 0x7acbea3b8ab7cdf4a595c6ed81e7d3e26038d494 inSync: {
  fetched: { amount: '349.478612562414524567 wstETH', duration: '1209ms' },
  offChain: { amount: '5463.196075047722536386 wstETH', duration: '16ms' }
}
```

any idea?

File: src/dex/ks-elastic/contract-math/ks-elastic-math.ts [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        let tick = TickList.getTick(tickList, Number(step.tickNext));
+        let liquidityNet = tick.liquidityNet;
+
+        liquidityNet = zeroForOne ? -liquidityNet : liquidityNet;
+        state.baseL = LiquidityMath.addDelta(state.baseL, liquidityNet);
+      }
+      state.tick = zeroForOne ? step.tickNext - 1n : step.tickNext;
+    }
+    if (isSell)
+      return BigInt.asUintN(
```
Comment (adedade) [inline src/dex/ks-elastic/contract-math/ks-elastic-math.ts:144]: Don't ask me how, but I was able to resolve most of the issues reported in https://github.com/paraswap/paraswap-dex-lib/pull/227#issuecomment-1347558181 by applying this hack

```
return BigInt.asUintN(
  256,
  -(state.amountSpecifiedRemaining == 0n
    ? state.amountCalculated
    : state.amountSpecifiedRemaining)
);
```

Comment (adedade) [reply src/dex/ks-elastic/contract-math/ks-elastic-math.ts:144]: You can see how most of the stuff works fine

```
[2022-12-12T22:47:36.732] [DEBUG] dex - kyberswapElastic 0xae338fe2521c609437ab996f187819ab59082f94 inSync: {
  fetched_: { amount: '0.000000980069788427 WETH', duration: '501ms' },
  offChain: { amount: '0.000000980069791754 WETH', duration: '0ms' }
}
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:47:51.035] [DEBUG] dex - kyberswapElastic 0xae338fe2521c609437ab996f187819ab59082f94 inSync: {
  fetched_: { amount: '0.000029402971461697 WETH', duration: '381ms' },
  offChain: { amount: '0.000029402974651063 WETH', duration: '1ms' }
}
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:49:24.791] [INFO] pool - Initial state loaded for pool 0x114de2affc6a335433dbe9d3d51a8f31a5591fdf
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:49:25.536] [DEBUG] dex - kyberswapElastic 0x114de2affc6a335433dbe9d3d51a8f31a5591fdf inSync: {
  fetched_: { amount: '0.016396720057456019 WETH', duration: '366ms' },
  offChain: { amount: '0.016396720163999999 WETH', duration: '1ms' }
}
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:49:39.219] [DEBUG] dex - kyberswapElastic 0xae338fe2521c609437ab996f187819ab59082f94 inSync: {
  fetched_: { amount: '0.000098009957211371 WETH', duration: '422ms' },
  offChain: { amount: '0.000098009990481065 WETH', duration: '0ms' }
```

but there might still be instances of 0 results e.g.

```
[2022-12-12T22:51:35.278] [INFO] pool - Initial state loaded for pool 0x7acbea3b8ab7cdf4a595c6ed81e7d3e26038d494
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:51:36.306] [DEBUG] dex - kyberswapElastic 0x7acbea3b8ab7cdf4a595c6ed81e7d3e26038d494 inSync: {
  fetched_: { amount: '365.088392203183658678 wstETH', duration: '664ms' },
  offChain: { amount: '0.0 wstETH', duration: '15ms' }
}
```

Comment (linhbkhn95) [reply src/dex/ks-elastic/contract-math/ks-elastic-math.ts:144]: can you share how to debug like you? It is the main key for me to solving all current problems.

Comment (adedade) [reply src/dex/ks-elastic/contract-math/ks-elastic-math.ts:144]: hey @linhbkhn95, I do calculate the off chain result via this snippet

```ts
export async function fetchQuote(
  tokenIn: string,
  tokenOut: string,
  amountIn: string,
  pool: string
): Promise<BigNumber> {
  const response = await axios.get<Quote>(
    'https://aggregator-api.kyberswap.com/arbitrum/route/encode',
    {
      params: {
        tokenIn: tokenIn,
        tokenOut: tokenOut,
        amountIn: amountIn,
        dexes: 'kyberswapv2',
        saveGas: '1',
        to: '0x0000000000000000000000000000000000000000',
      },
    }
  );

  const quote = response.data;

  // We need to find the swap that does only use the Pool that we are trying to get a quote for
  const poolSwap = quote.swaps.find((swaps) => {
    return swaps.length === 1 && isSameAddress(pool, swaps[0].pool);
  });

  if (poolSwap === undefined) {
    throw new Error('Unable to find quote for pool: ' + pool);
  }

  const { amountOut } = poolSwap[0];
  return BigNumber.from(amountOut);
}
```

Comment (linhbkhn95) [reply src/dex/ks-elastic/contract-math/ks-elastic-math.ts]: can you provide the full code or ref from where?

Comment (adedade) [reply src/dex/ks-elastic/contract-math/ks-elastic-math.ts]: hey @linhbkhn95 I do run `fetchQuote` (my implementation above) and `getOutputAmountProMM` (your implementation) whenever a swap happen on a pool, then print out the two amounts. I am happy to provide the full snippet on Discord, my username is `adedade#3981` add me :)

Comment (linhbkhn95) [reply src/dex/ks-elastic/contract-math/ks-elastic-math.ts]: Thanks, I sent my invitation to add friend

**Pull request closed without merging**